### DISCUSSION TITLE
Curious: MacOS Has Smooth Transitions

### DISCUSSION THREAD

**Discussion by anonymousaga** [Dansoftowner/jSystemThemeDetector#15] [Q&A]
MacOS, being extremely polished by Apple, has smooth transitions that fade from dark to light in apps (such as Finder, Notes, etc.). What surprises me is that even my JAR app is using this fade transition. I never programmed this fade transition, so how does it work? On Windows, the app changes color instantly, like I expected.

P.S. I love this polish on MacOS and how it fits in with other apps, I am just curious how the java app itself is changing its colors with a fading transition.

Update: I'm using AWT/Swing

**Comment by Dansoftowner** [comment]
Wow, that's interesting!
Are you using JavaFX?

**Reply by anonymousaga** [reply]
Nope, I'm using AWT/Swing.

**Reply by Dansoftowner** [reply]
Then I think that the fade transition is implemented in Swing itself on MacOS.

**Reply by anonymousaga** [reply]
I wrote a tiny program including these lines after the GUI was initialized (using wait method)

```
wait(3000);
panel.setBackground(Color.decode("#cce6ff"));
```

And it didn't fade. Instead, it switched instantly. So I don't think Swing is using the faded transitions.

**Reply by Dansoftowner** [reply]
Hmmm, even if you use the `UIManager` for changing the look and feel?

**Reply by anonymousaga** [reply]
Oh, I never used UIManager. I just changed the background of my panes and the text color.

    ```
    wait(3000);
        try {
            // Set cross-platform Java L&F (also called "Metal")
            UIManager.setLookAndFeel(UIManager.getCrossPlatformLookAndFeelClassName());
        } catch (UnsupportedLookAndFeelException e) {
            // handle exception
        } catch (ClassNotFoundException e) {
            // handle exception
        } catch (InstantiationException e) {
            // handle exception
        } catch (IllegalAccessException e) {
            // handle exception
        } catch (Exception e) {
       }

```
Did nothing, as I only used that command before initializing a shape. Sorry if I'm making silly mistakes, Im a beginner and I started java 2 months ago

**Reply by Dansoftowner** [reply]
Well, then I have no idea why this fade-transition happens but it's really interesting!
I still think It's somehow related to the Swing/AWT framework, but I'm not a Swing expert so I can't tell you the exact reason why this is happening.

**Reply by anonymousaga** [reply]
Yeah, it's really interesting. Should I leave this discussion open if anybody else thinks of why this is happening?

**Reply by Dansoftowner** [reply]
> I wrote a tiny program including these lines after the GUI was initialized (using wait method)
> 
> ```
> wait(3000);
> panel.setBackground(Color.decode("#cce6ff"));
> ```
> 
> And it didn't fade. Instead, it switched instantly. So I don't think Swing is using the faded transitions.

BTW: do you execute these lines on a background thread?

**Reply by anonymousaga** [reply]
The entire GUI is in a thread, so yes. But the Dark Mode listener is also in the GUI thread (though run-later creates its own thread).

That's how it works in my original program so I tried to imitate it in my test program

**Reply by Dansoftowner** [reply]
If yes, you should wrap the `panel.setBackground(...)` part into `SwingUtilities.invokeLater` to execute it on the UI thread.
Like this: 
```java
new Thread(() -> {
     wait(3000);
     SwingUtilities.invokeLater(() -> panel.setBackground(Color.decode("#cce6ff")));
}).start();
```

**Reply by Dansoftowner** [reply]
> The entire GUI is in a thread, so yes. But the Dark Mode listener is also in the GUI thread (though run-later creates its own thread).
> 
> That's how it works in my original program so I tried to imitate it in my test program

Well, if you don't start a background thread, that means you freeze the entire UI thread for 3000 milliseconds ( so the transition might not work because of that).
You should start a separate thread for waiting (as I showed you above).

**Reply by Dansoftowner** [reply]
The dark mode listener is a little bit different, I don't want to get into that.

**Reply by anonymousaga** [reply]
I tried that, but it is still instant. This is how it looks with the thread: https://www.dropbox.com/t/DMvyoCQBnO9Tn4AD
this is how it looks when changing light to dark mode: https://www.dropbox.com/t/uuyIkh9gtkhBX28Q

**Reply by Dansoftowner** [reply]
> Yeah, it's really interesting. Should I leave this discussion open if anybody else thinks of why this is happening?

Okay, then we should leave this discussion open! :)

**Comment by nsmoker** [comment]
This happens in JavaFX as well, it's really nice, but I too have no clue why it would be happening.
